src/lib/support/tests/TestUtf8.cpp

```diff
@@ -0,0 +1,151 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <functional>
+
+#include <lib/support/UnitTestRegistration.h>
+#include <lib/support/utf8.h>
+
+#include <nlunit-test.h>
+
+namespace {
+
+using namespace chip;
+
+bool IsValidCStringAsUtf8(const char * s)
+{
+    return Utf8::IsValid(CharSpan(s, strlen(s)));
+}
+
+void TestValidStrings(nlTestSuite * inSuite, void * inContext)
+{
+    NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan())); // empty span ok
+
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8(""));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("abc"));
+
+    // Various tests from https://www.w3.org/2001/06/utf-8-wrong/UTF-8-test.html
+
+    // Generic UTF8
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("κόσμε"));
+
+    // First possible sequence of a certain length
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8(""));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("ࠀ"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("𐀀"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("�����"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("������"));
+
+    // Last possible sequence of a certain length
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8(""));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("߿"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("￿"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("����"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("�����"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("������"));
+
+    // Other boundary conditions
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("퟿"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8(""));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("�"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("􏿿"));
+    NL_TEST_ASSERT(inSuite, IsValidCStringAsUtf8("����"));
+}
+
+#define TEST_INVALID_BYTES(...)                                                                                                    \
+    {                                                                                                                              \
+        uint8_t _buff[] = { __VA_ARGS__ };                                                                                         \
+        CharSpan _span(reinterpret_cast<const char *>(_buff), sizeof(_buff));                                                      \
+        NL_TEST_ASSERT(inSuite, !Utf8::IsValid(_span));                                                                            \
+    }                                                                                                                              \
+    (void) 0
+
+void TestInvalidStrings(nlTestSuite * inSuite, void * inContext)
+{
+    // cannot embed zeroes
```

Comment:
Embedding zeros is legal, let's remove this test.

Reply:
Disagree for CHIP. This is a requirement I want.